File: backend/src/services/llm/LLMServiceFactory.ts
/**
 * LLM Service Factory
 * Creates the appropriate LLM service based on configuration
 * Currently supports Claude Code, extensible for future providers
 */

import type { LLMService, LLMConfig, LLMProvider } from './LLMService.js';
import { ClaudeCodeService } from './ClaudeCodeService.js';
import { DEFAULT_LLM_CONFIG } from './LLMService.js';

/**
 * Factory for creating LLM service instances
 */
export class LLMServiceFactory {
  private static instance: LLMService | null = null;
  private static currentProvider: LLMProvider | null = null;

  /**
   * Create an LLM service based on the provided configuration
   * @param config LLM configuration
   * @returns LLMService instance
   */
  static create(config: LLMConfig = DEFAULT_LLM_CONFIG): LLMService {
    switch (config.provider) {
      case 'claude-code':
        return new ClaudeCodeService();

      // case 'claude-api':
      //   // Future: Implement ClaudeAPIService
      //   throw new Error(
      //     'Claude API provider is not yet implemented. Please use claude-code for now.'
      //   );
      //
      // case 'openai':
      //   // Future: Implement OpenAIService
      //   throw new Error(
      //     'OpenAI provider is not yet implemented. Please use claude-code for now.'
      //   );

      default:
        throw new Error(`Unknown LLM provider: ${config.provider}`);
    }
  }

  /**
   * Get or create a singleton LLM service instance
   * Recreates if provider changes
   * @param config LLM configuration
   * @returns LLMService instance
   */
  static getInstance(config: LLMConfig = DEFAULT_LLM_CONFIG): LLMService {
    // Recreate if provider changed
    if (this.currentProvider !== config.provider) {
      this.instance = null;
    }

    if (!this.instance) {
      this.instance = this.create(config);
      this.currentProvider = config.provider;
    }

    return this.instance;
  }

  /**
   * Clear the singleton instance
   * Useful for testing or when config changes
   */
  static clearInstance(): void {
    this.instance = null;
    this.currentProvider = null;
  }

  /**
   * Get list of available providers
   */
  static getAvailableProviders(): LLMProvider[] {
    return ['claude-code']; // Add more as they're implemented
  }

  /**
   * Check if a provider is available (implemented)
   */
  static isProviderAvailable(provider: LLMProvider): boolean {
    return this.getAvailableProviders().includes(provider);
  }
}

File: src/main/kotlin/com/ankideku/data/remote/llm/LlmServiceFactory.kt
package com.ankideku.data.remote.llm

/**
 * Factory for creating LLM service instances.
 * Creates the appropriate service based on configuration.
 */
object LlmServiceFactory {
    private var instance: LlmService? = null
    private var currentProvider: LlmProvider? = null

    /**
     * Create an LLM service based on the provided configuration.
     */
    fun create(config: LlmConfig = DEFAULT_LLM_CONFIG): LlmService {
        return when (config.provider) {
            LlmProvider.CLAUDE_CODE -> ClaudeCliService()
            LlmProvider.MOCK -> MockLlmService()
        }
    }

    /**
     * Get or create a singleton LLM service instance.
     * Recreates if provider changes.
     */
    fun getInstance(config: LlmConfig = DEFAULT_LLM_CONFIG): LlmService {
        // Recreate if provider changed
        if (currentProvider != config.provider) {
            instance = null
        }

        if (instance == null) {
            instance = create(config)
            currentProvider = config.provider
        }

        return instance!!
    }

    /**
     * Clear the singleton instance.
     * Useful for testing or when config changes.
     */
    fun clearInstance() {
        instance = null
        currentProvider = null
    }

    /**
     * Get list of available providers.
     */
    fun getAvailableProviders(): List<LlmProvider> {
        return LlmProvider.entries
    }

    /**
     * Check if a provider is available (implemented).
     */
    fun isProviderAvailable(provider: LlmProvider): Boolean {
        return provider in getAvailableProviders()
    }
}
